L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,425 @@
+1. Customizable verification of the peer certificate: perform custom validation checks on the peer's certificate chain after it has been verified to chain up to a root of trust, or fully customize path building and cryptographic verification.
+1. TLS version configuration: optionally configure the minimum and maximum of the TLS versions to be negotiated during the handshake.
+1. Support for certificate revocation via Certificate Revocation Lists (CRLs).
+
+## Background
+
+The current TLS implementation (the SSL credentials API) in gRPC C core has some restrictions:
+1. Lack of the credential reloading support. Once an application starts, it can never reload its private key, certificate chain conveying the application's identity, and root certificates without shutting down.
+1. Inflexibility of how the peer certificate chain is verified. There is no API to add additional custom verification checks.
+1. Inability to choose the maximum/minimum TLS version to be negotiated during the TLS handshake.
+1. No support for certificate revocation.
+
+We propose an API that supports the above features and meets the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones do not fit their needs.
+1. The credential reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection.
+1. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API.
+
+### Related Proposals:
+
+1. [L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98) - this was subsumed into this proposal.
+1. [A69: Crl Providers](https://github.com/grpc/proposal/pull/382)
+
+## Proposal
+
+### TLS credentials and TLS Options
+This part of the proposal introduces the base options class, called `TlsCredentialsOptions`, that has a large set of options for configuring TLS connections. There are 2 derived classes of `TlsCredentialsOptions`, called `TlsChannelCredentialsOptions` and `TlsServerCredentialsOptions`, that users create and can use to build channel credentials or server credentials instances, respectively. The API for building these channel/server credentials instances are called `TlsCredentials`/`TlsServerCredentials` respectively.
+
+```c++
+// Base class of configurable options specified by users to configure their
+// certain security features supported in TLS.
+class TlsCredentialsOptions {
+ public:
+  ~TlsCredentialsOptions();
+
+  // Copy constructor does a deep copy. No assignment permitted.
+  TlsCredentialsOptions(const TlsCredentialsOptions& other);
+
+  TlsCredentialsOptions& operator=(const TlsCredentialsOptions& other) = delete;
+
+  // ---- Setters for member fields ----
+  // Sets the certificate provider used to store root certificates and the
```

Comment:
micro-nit: I think this sentence could benefit from a couple Oxford commas for clarity. For example:

"""
Sets the certificate provider used to store root certificates, the certificate chain conveying the application's identity, and the corresponding private key.
"""

Reply:
This is an interesting case because it's not {1, 2, and 3}, but rather
It can set {root certificates} and {the cert chain and corresponding private key}
WDYT?

Reply:
Fair point. To convey that subtlety, WDYT about:

// Sets the certificate provider, which is used to store:
// 1. The root certificates used to (cryptographically) verify peer certificate chains.
// 2. The certificate chain conveying the application's identity and the corresponding private key.

WDYT?

Reply:
I like it, changed